### ISSUE TITLE
`butane` + async runtime

### ISSUE THREAD

**Issue by TmLev** [Electron100/butane#13]
Is it possible to use `butane` with asynchronous runtime like, for example, `tokio`?

**Comment by Electron100** [comment]
Not directly, aside from using something like `tokio::task::spawn_blocking` of course. I'm not opposed to some deeper async integration, but it's not something that I'm expecting to implement soon

**Comment by Electron100** [comment]
Closing this out as it's not something I'm expecting to do any time soon. I'd be open to a pull request for this, although it would likely only be applicable to some backends, since e.g. sqlite is always going to be based on libsqlite which isn't async-aware.

**Comment by ibraheemdev** [comment]
The sqlite backend would still have to expose async methods that do something like `spawn_blocking`.

**Comment by Electron100** [comment]
@ibraheemdev sure, but the consumer of Butane could just as well make the spawn_blocking calls themselves. That's the same approach taken with Diesel, as in this Actix Web example https://actix.rs/docs/databases/.

**Comment by ibraheemdev** [comment]
Yeah, I was just saying that if you wanted to expose an async API, ideally everything should be usable directly in an async context.

**Comment by Electron100** [comment]
Reactivating as this is under active consideration

**Comment by jayvdb** [comment]
I've merged master into the latest async branch , disabling `r2d2` feature.
https://github.com/jayvdb/butane/tree/async-no-r2d2-jv-merge-master

**Comment by Electron100** [comment]
@jayvdb remind me what is the intended difference between that branch and `async-wip`? That also has r2d2 currently disabled.

I've just added [async_checklist.md](https://github.com/Electron100/butane/blob/async-wip/async_checklist.md ) to attempt to track the remaining work on that branch.

**Comment by jayvdb** [comment]
I think I removed r2r2 from the features list so that I could use `--all-features`, and it has a recent `master` merged into it.

**Comment by jayvdb** [comment]
https://github.com/nvksv/maybe-async-cfg/pull/4 has been merged and released, which we can use on the butane branch to reduce the dependencies.

**Comment by Electron100** [comment]
Hopefully better late than never, as of this week, I *believe* the [async-wip](https://github.com/Electron100/butane/tree/async-wip) branch is finally functional and API-stable (although that's not a guarantee). Remaining tasks are tracked in https://github.com/Electron100/butane/blob/async-wip/async_checklist.md. The biggest outstanding items are to properly prove/ensure soundness in some unsafe code used to bridge between sync and async when necessary and to provide support for async connections pools (deadpool and/or bb8, although this is not blocking as it could also be done out of tree).

If anybody is interested in giving it a whirl, feedback is welcome.

**Comment by RobertoMaurizzi** [comment]
About deadpool (and sorry if this is obviously wrong/impossible :sweat_smile: ) but wouldn't https://crates.io/crates/deadpool-postgres be usable if we can give it a `tokio_postgres::Config`?
However, I just started checking the code but I can't see any use of that Config object, so maybe not...

**Comment by Electron100** [comment]
Butane doesn't expose its Postgres implementation. From a quick scan of the
deadpool docs, the way to do deadpool integration would be to create a type
like `butane::db::deadpool::ConnectionManager` which is constructed with a
butane `ConnectionSpec` and implements   Manager in deadpool::managed - Rust
<https://docs.rs/deadpool/latest/deadpool/managed/trait.Manager.html> to
create a new connection from that spec on demand. Should be straightforward
but I'm traveling for work this week so unlikely to get to it until next
week at the earliest.

**Comment by jayvdb** [comment]
@Electron100 , thanks for finishing that off. IMO it would be good to push the async branch to master.  Squashed under your id would be my pref as the history is quite messy?  You've done all the hard bits.

Then the remaining TODOs can be done by separate people as normal PRs.

Maybe include a note at the top of the README that `master` is unstable atm while we finish the async work.

**Comment by Electron100** [comment]
Good point. I've created the PR. I'd like to give it a once over myself over the next few days (and of course other reviews are very welcome) to look for anything egregiously bad/incomplete before merging, but will plan to merge it within the week.

**Comment by Electron100** [comment]
PR is in master now!
I'm waiting to close this issue fully until we've done some of the tracked cleanup.

**Comment by RobertoMaurizzi** [comment]
Unfortunately I was not well during most of the holidays (as usual, I might add :sweat_smile: )
I was able yesterday to go over my Leptos application to try to get deadpool working, however I'm stuck because I couldn't find how to go from a pool to a valid database connection. As you wrote in the README.md,  the `ConnectionManager` doesn't iimplement `ConnectionMethods` anymore... but if I `deref_mut()` a `ConnectionAsync`, I can call `setup_db_async` but I can't use it for a `query!().load()`...
What am I missing? :sweat_smile:

**Comment by jayvdb** [comment]
You will want `ConnectionMethodsAsync` instead.
Also import `butane::ManyOpsAsync` & `butane::prelude_async::ForeignKeyOpsAsync` to be able to `.load()` on those objects, and sometimes import `butane::DataObjectOpsAsync` is needed.
And some of the functions/macros now have an `_async` suffix, like `butane::db::connect_async` and `butane::find_async`.

**Comment by RobertoMaurizzi** [comment]
I'm doing:

```rust
let pool = deadpool::managed::Pool::builder(manager).build().unwrap();
let mut conn: deadpool::managed::Object<ConnectionManager> = pool.get().await.unwrap().deref_mut();
```
but `conn` is a `&mut ConnectionAsync` instead of a `ConnectionMethodsAsync`.
I enabled the features in my app as  `butane/pg`, `butane/async` (even if it shouldn't be necessary) and `butane/deadpool`.
:zany_face:

**Comment by Electron100** [comment]
@RobertoMaurizzi `ConnectionAsync` impls `ConnectionMethodsAsync` though, so it should work. Can you show the line that fails?

**Comment by RobertoMaurizzi** [comment]
I end up with a bunch of
```
let matching_users = query!(User, username == { username.as_str() }).load(&dbconn).await?;
                                                                     ^^^^ the trait `pg::PgConnectionLike` is not implemented for `&mut ConnectionAsync`
```
Surprisingly enough, a `find_async` seems to work (I don't get any compiler error) but `query!` does not (and there's no `query_async`, it seems). I was under the impresion that importing the `prelude::async` instead of `prelude` would call in all the code for async usage... maybe I'm still importing something that overrides that? :thinking:

**Comment by Electron100** [comment]
I think something must be wrong with your imports. Any way you're able to share a more complete sample of the code that's failing?

I updated deadpool_test_pg_async with a line very similar to what you show above and it works just fine. You can see it here https://github.com/Electron100/butane/compare/master...deadpool_test

**Comment by RobertoMaurizzi** [comment]
To avoid all the very messy setup of my Leptos application (that I'll likely port to Leptos 0.7 by rewriting it as a single-crate instead of the quite confusing multi crate approach I tried before...) I tried to add pool support to my first butane test program, one that reads Stable Diffusion image data and saves them in a database...

At first it was exactly the same, then while thinking why and how the test works and my code doesn't, I noticed my connection was a `&&ConnectionAsync`! :facepalm: because that's what it was before adding pool support.

I guess the compiler started protesting about non-implemented traits because, well, the parameter IS a reference as required (but to another reference) and the unimplemented trait exists for the non-referenced type BUT not for the reference itself... I didn't realize that deref_mut() gave me a reference already and the existing code was already using `&dbconn`, so it happened for both projects :sweat_smile: 

I do find the error message extremely misleading: 
```
the trait `pg::PgConnectionLike` is not implemented for `&mut ConnectionAsync`
help: the trait `pg::PgConnectionLike` is implemented for `PgConnection`
```
doesn't really make me think the problem is that I have a reference to the right value instead of the value itself...

If I can bring the image-data-reading example code up to a decent state (like.. actually using async to read the images!) I'll send you a PR to add it to the examples.

Thanks for your help! :smile:

**Comment by Electron100** [comment]
We haven't had any breaking changes in a while now, so I'd like to go ahead and release the next version of Butane to crates.io. @jayvdb or @RobertoMaurizzi let me know if you have any reason I should hold.

**Comment by RobertoMaurizzi** [comment]
I'm using a local git clone pointing at master so it's not a problem for me :D 
I was thinking it might be a good idea to improve the docs with commonly used things (if already available)...
From the top of my head from what I've struggled with up to now:
- how to access the inner object of an AutoPk
- how to assign a ForeignKey having the naked pk
- how to count how many record are in a query

@jayvdb @dhbradshaw do you have any other suggestion?

**Comment by Electron100** [comment]
@RobertoMaurizzi good point on the first one, this seems super painful today. I added a `Deref` impl in #302. Does that look better?

For "how to assign a ForeignKey having the naked pk", I just check the docs and they say "Initialize using `From` or `from_pk`". The latter is what you want in your case. Any suggestions on how to reword to make this any clearer?

For the last one, are you saying you want to count records without actually retrieving them?

**Comment by RobertoMaurizzi** [comment]
I was thinking more about including the most obvious things in the getting started document.
Maybe it's only me but I do find it much more easy to use than RustDoc, in addition it's super useful to get a feeling of how Butane "looks and feels" for typical database operations as they're used when building API endpoints/Leptos server functions.

**Comment by RobertoMaurizzi** [comment]
Another possible useful addition to the getting started: how to manage an optional foreign key, like:
```rust
    let registered_webpush_clients = query!(
        WebPushSubscriptionData,
        user_id == { Some(ForeignKey::<User>::from_pk(user.id)) }
    )
```
The error details are literally telling you to use `ForeignKey::<Model>::from_pk`... but most editors hide it :-/
(in this specific case I don't need the Option a delete cascade would work better, I only added it due to my experiments on how to find the database structure to try to implement an admin-like feature in this application).

**Comment by Electron100** [comment]
Closing this as async support has been merged for a while and 0.8 is now on crates.io
@RobertoMaurizzi I've taken a couple of your suggestions but likely missed others. Please feel free to raise a PR improving the docs, or a separate issue dedicated to discussing the improvements you're suggesting.
